File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode* result_head = new ListNode();

        ListNode* result_cur = result_head;
        while(l1 != nullptr && l2 != nullptr){
            int sum = l1->val + l2->val;
            l1 = l1->next; l2 = l2->next;

            sum += carry;
            carry = sum / 10;
            result_cur->next = new ListNode(sum % 10);
            result_cur = result_cur->next;
        }
        while(l1 != nullptr){
            int sum = l1->val + carry;
            l1= l1->next;
            
            carry = sum / 10;
            result_cur->next = new ListNode(sum % 10);
            result_cur = result_cur->next;
        }
        while(l2 != nullptr){
            int sum = l2->val + carry;
            l2= l2->next;

            carry = sum / 10;
            result_cur->next = new ListNode(sum % 10);
            result_cur = result_cur->next;
        }

        if(carry) result_cur->next = new ListNode(carry);

        return result_head->next;
    }
};

File: 0002-add-two-numbers/0002-add-two-numbers.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val sumHead = ListNode(-1)
        var cur = sumHead

        var head1 = l1
        var head2 = l2
        var carry = 0
        while(head1 != null || head2 != null){
            var sum = (head1?.`val` ?: 0) + (head2?.`val` ?: 0) + carry
            head1 = head1?.next
            head2 = head2?.next

            if(sum >= 10){
                sum -= 10
                carry = 1
            }else{
                carry = 0
            }
            cur.next = ListNode(sum)
            cur = cur.next
        }
        if(carry != 0) cur.next = ListNode(1)

        return sumHead.next
    }
}